Detectors/ITSMFT/ITS/workflow/src/CalibratorSpec.cxx

```diff
@@ -0,0 +1,1083 @@
+        if (charge > this->max || charge < this->min) {
+          LOG(error) << "charge value " << charge << " out of range for min " << this->min
+                     << " and max " << this->max << " (range: " << *(this->nRange) << ")";
+          exit(1);
+        }
+        // LOG(info) << "before";
+        this->pixelHits[chipID][col][charge - this->min]++;
+
+      } // for (digits)
+
+    } // for (RUs)
+
+  } // for (ROFs)
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Calculate the average threshold given a vector of threshold objects
+template <class Mapping>
+void ITSCalibrator<Mapping>::find_average(const std::vector<threshold_obj>& data,
+                                          float& avg, float& rms)
+{
+
+  float sum = 0;
+  unsigned int counts = 0;
+  for (const threshold_obj& t : data) {
+    if (t.success) {
+      sum += t.threshold;
+      counts++;
+    }
+  }
+
+  avg = (!counts) ? 0. : sum / (float)counts;
+  sum = 0.;
+  for (const threshold_obj& t : data) {
+    if (t.success) {
+      sum += (avg - (float)t.threshold) * (avg - (float)t.threshold);
+    }
+  }
+  rms = (!counts) ? 0. : std::sqrt(sum / (float)counts);
+
+  return;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+template <class Mapping>
+void ITSCalibrator<Mapping>::add_db_entry(const short int& chipID, const std::string* name,
+                                          const short int& avg, const float& rms, bool status, o2::dcs::DCSconfigObject_t& tuning)
+{
+
+  // Obtain specific chip information from the chip ID (layer, stave, ...)
+  int lay, sta, ssta, mod, chipInMod; // layer, stave, sub stave, module, chip
+  this->mp.expandChipInfoHW(chipID, lay, sta, ssta, mod, chipInMod);
+
+  std::string stave = 'L' + std::to_string(lay) + '_' + std::to_string(sta);
+
+  o2::dcs::addConfigItem(tuning, "Stave", stave);
+  o2::dcs::addConfigItem(tuning, "Hs_pos", std::to_string(ssta));
+  o2::dcs::addConfigItem(tuning, "Hic_Pos", std::to_string(mod));
+  o2::dcs::addConfigItem(tuning, "ChipID", std::to_string(chipInMod));
+  o2::dcs::addConfigItem(tuning, *name, std::to_string(avg));
+  o2::dcs::addConfigItem(tuning, "Rms", std::to_string(rms));
+  o2::dcs::addConfigItem(tuning, "Status", std::to_string(status)); // pass or fail
+}
+
+//////////////////////////////////////////////////////////////////////////////
+template <class Mapping>
+void ITSCalibrator<Mapping>::send_to_ccdb(std::string* name,
+                                          o2::dcs::DCSconfigObject_t& tuning, EndOfStreamContext& ec)
+{
+
+  // Below is CCDB stuff
+  long tstart, tend;
+  tstart = o2::ccdb::getCurrentTimestamp();
+  constexpr long SECONDSPERYEAR = 365 * 24 * 60 * 60;
+  tend = o2::ccdb::getFutureTimestamp(SECONDSPERYEAR);
+
+  auto class_name = o2::utils::MemFileHelper::getClassName(tuning);
+
+  // Create metadata for database object
+  std::string* ft;
```

Comment:
why pointer rather than object?

Reply:
A pointer is better in this case so that the object is not initialized twice -- though in theory either would work. I don't have a strong preference one way or the other.

Reply:
`std::string s;` declaration does not involve any heap memory allocation (in opposite to `new std::string()` which does 2 heap allocations: one to create the string object and another one to set the string value if it is longer than the fixed size internal buffer, but this is STL implementation dependent).

Reply:
`std::string s` calls the default constructor for `std::string` and creates a new (empty) `std::string` which is then stored on the stack. `s = std::string("stuff")` calls a copy constructor for `std::string` and creates a second `std::string` object, which replaces the original (empty) `std::string`. The deconstructor should then called for the original (empty) `std::string`.

To my understanding whether or not the object is stored on the heap or stack is not so important, but using a pointer the object is at least only created once.

Reply:
If really one is worried about efficiency here the best is to use `char*` but I think really this will have no perceivable affect at all on the code...

Reply:
this is just in our coding guidelines: avoid bare `new` operators if possible.

Reply:
OK, understood -- thanks. I will change to char array.

Reply:
ok, it becomes somewhat academic, but... what is expensive the heap memory allocation, creation of the stack object is very cheap. The 'std::string x;' creates just a fixed size stack object, and `x = "something"`; will call assignment operator with just 1 memory allocation.
With the `new std::string("something")`, you have 2 allocations.
That said, both cases will be likely optimized out to predefined const strings with needed one picked depending on the if condition. What matters is that the code with local objects is safer than that will explicit new / deletes.